File: backtracking/46.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2019-11-19 13:17:20
'''
import sys
import os

sys.path.append("..")


class Solution(object):
    def permute(self, nums):
        result = []
        self.get_permute(result, nums, [])
        return result

    def get_permute(self, result, nums, tmp):
        if(len(nums) == 0): result.append(tmp[:])

        for i in range(len(nums)):
            tmp.append(nums[i])
            self.get_permute(result, nums[:i] + nums[i+1:], tmp)
            tmp.pop()


class Solution2(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # insert solution

        last_re, current_re = [[]], []

        for i in range(len(nums)):
            for _re in last_re:
                length_j = len(_re)
                for j in range(length_j+1):
                    tmp = _re[:j] + [nums[i]] + _re[j:]
                    current_re.append(tmp)
            last_re = current_re
            current_re = []

        return last_re

ss = Solution()
print(ss.permute([]))
                        

File: backtracking/46.cpp
/*
 * @Author: httermin
 * @Date: 2019-11-18 16:48:18
 */

#include<string>
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class Solution {
    //backtracking
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> tmp;
        vector<bool> isValid(nums.size(), true);
        getPermute(nums, result, tmp, isValid);
        return result;
    }

    void getPermute(vector<int>& nums, vector<vector<int>>& result, vector<int>& tmp, vector<bool> isValid){
        if(tmp.size() == nums.size()) result.push_back(tmp);
        
        for(int i=0; i<nums.size(); i++){
            if(isValid[i]){
                isValid[i] = false;
                tmp.push_back(nums[i]);
                getPermute(nums, result, tmp, isValid);
                isValid[i] = true;
                tmp.pop_back();
            }
        }
        
    }
};

int main(){
    int a[] = {1, 2, 3};
    vector<int> candidates(a, a+3);

    int n=5;
    Solution ss;

    vector<vector<int>> result = ss.permute(candidates);
    for(int i=0; i<result.size(); i++){
        for(int j=0; j<result[i].size(); j++){
            cout<<result[i][j]<<" ";
        }
        cout<<" "<<endl;
    }

    return 1;
}